statev2/state/src/applicator/order_book.rs

```diff
@@ -0,0 +1,488 @@
+        );
+        Ok(())
+    }
+
+    /// Add a validity proof for an order
+    pub fn add_order_validity_proof(&self, msg: AddOrderValidityProofMsg) -> Result<()> {
+        // Deserialize the proof bundle
+        let id: OrderIdentifier = msg
+            .order_id
+            .unwrap_or_default()
+            .try_into()
+            .map_err(|e: UuidError| StateApplicatorError::Parse(e.to_string()))?;
+
+        let bundle: OrderValidityProofBundle = serde_json::from_slice(&msg.proof)
+            .map_err(|e| StateApplicatorError::Parse(e.to_string()))?;
+
+        let tx = self
+            .db()
+            .new_write_tx()
+            .map_err(StateApplicatorError::Storage)?;
+        Self::attach_validity_proof_with_tx(&id, bundle, &tx)?;
+        let order_info = Self::read_order_info(&id, &tx)?;
+        tx.commit().map_err(StateApplicatorError::Storage)?;
+
+        self.system_bus().publish(
+            ORDER_STATE_CHANGE_TOPIC.to_string(),
+            SystemBusMessage::OrderStateChange { order: order_info },
+        );
+        Ok(())
+    }
+
+    /// Nullify orders indexed by a given wallet share nullifier
+    pub fn nullify_orders(&self, msg: NullifyOrdersMsg) -> Result<()> {
+        let nullifier: Scalar = msg.nullifier.unwrap_or_default().into();
+        let tx = self
+            .db()
+            .new_write_tx()
+            .map_err(StateApplicatorError::Storage)?;
+
+        self.nullify_orders_with_tx(nullifier, &tx)?;
+        tx.commit().map_err(StateApplicatorError::Storage)
+    }
+
+    // ------------------------
+    // | Order Update Helpers |
+    // ------------------------
+
+    /// Add an order to the book
+    ///
+    /// TODO: For an initial implementation we do not re-index based on local orders or
+    /// verified orders. This will be added with the getter implementations
+    fn add_order_with_tx(order: &NetworkOrder, tx: &DbTxn<'_, RW>) -> Result<()> {
+        // Write the order to storage
+        Self::write_order_info(order, tx)?;
+        // Add the order to the set of orders indexed by the nullifier
+        Self::append_to_nullifier_set(order.public_share_nullifier, order.id, tx)
+    }
+
+    /// Update the validity proof for an order
+    fn attach_validity_proof_with_tx(
+        order_id: &OrderIdentifier,
+        proof: OrderValidityProofBundle,
+        tx: &DbTxn<'_, RW>,
+    ) -> Result<()> {
+        // Read the order's current info
+        let mut order_info = Self::read_order_info(order_id, tx)?;
+
+        // Re-index based on the proof's nullifier
+        let prev_nullifier = order_info.public_share_nullifier;
+        let new_nullifier = proof.reblind_proof.statement.original_shares_nullifier;
+        if prev_nullifier != new_nullifier {
+            Self::update_order_nullifier(order_id, prev_nullifier, new_nullifier, tx)?;
+        }
+
+        // Update the order's info
+        order_info.state = NetworkOrderState::Verified;
```

Comment:
Does this mean `attach_validity_proof_with_tx` is assumed to be called after the validity proof has been verified? I don't see that happening in `add_order_validity_proof`

Reply:
Yes that's correct, this assumption isn't mandated by the interface (I see no simple way to do so), but instead follows the pattern that the rest of the relayer lays out: authentication happens at the network and gossip layers which then dispatch to the storage layer